### ISSUE TITLE
Storage of tour assets - discussion

### ISSUE THREAD

**Issue by jrosindell** [OneZoom/OZtree#601]
I'm opening this because I have a concern about keeping the tour content reliably available into the future if tour assets are always stored by third parties.  This is especially relevant in case we invite third parties to create tours in future.  My worry is that we'll be left with tours that are full of hard coded links to resources that no longer exist or worse that the rights of use associated with those resources will change and there will be no paper trail to prove that we have those assets to use in tours based on a previous license of those assets. The tour creators might not think of this and might not be committed to maintaining the links, or might not themselves be in control of where those links go to. Whilst we can't guarantee tour quality if it's a third party product, it will definitely reflect badly on us if things look absolutely broken.

I was therefore wondering if for images and audio only we store the material on our own server - then for videos we use a service like Vimeo.  Under that model I'd propose that we keep the number of videos quite small to avoid future problems - they are really an accent rather than the main event - after all we want the tree and built in animations to play a major role and not for these to be reduced to a playlist of videos that are linked somehow.  I'd also suggest that we don't allow third parties the rights to create tours with videos automatically - they would have to discuss that with us and maybe make the video available via our own Vimeo hosting so that we can guarantee long-term availability. When third parties create tours they would have to upload the image and auto assets needed and provide information on the rights of use for those which can then be stored.

So, the purpose of this issue is to open a discussion about these concerns and to find out from @lentinj how feasible it would be to add a bare bones system for storing tours media on our own servers.

**Comment by lentinj** [comment]
Link-rot and asset licencing are different issues I think.

For link-rot, one pretty easy option for images is to inline them into the tour JSON as [data: URIs](https://en.wikipedia.org/wiki/Data_URI_scheme). This isn't the most efficient option, as we end up storing the image data in the tour table, which won't be ideal for schema performance. However, everything would just-work as we expect, and a tour would be represented as a single file, which is pretty useful.

Another option would be to rewrite tours on load to pre-cache images on load. When rendering JSON -> HTML, inspect URLs and instead of blindly inserting into HTML, fetch the content, dump somewhere web accessible under the SHA256 of the URL if not already there, and return a URL to that. This means that we can cache resources without worrying about copying the cache when manhandling database dumps (since it'll re-populate itself, assuming the resources are still available). Downside is if a resource gets updated, we won't update our copy without modifying the URL of the original asset somehow.

As for licencing, how do you prove you had the rights to start off with, ignoring changes down the line? Storing an attribution URL alongside the image is probably the only generic, concrete thing you could do. Trying to store the contents of this page to act as evidence that it used to be CC sounds more painful than it's worth though.

**Comment by lentinj** [comment]
It might also be worth insisting that images, instead of being arbitrary URLs, are higher level constructs, e.g:

* ``oz:(img_src):(img_src_id)``
* ``wikimedia:Western_Lowland_Gorilla_Feb09_02.jpg``

This has a few advantages:

* Gatekeeping new sources gives us a chance to ensure that images are allowed to appear on OneZoom, and/or how to extract attribution to ensure this is the case.
* Attribution can be plumbed in automatically, rather than relying on the tour creator to do the legwork
* It makes the moderation problem slightly more manageable. Whilst I'm sure Wikimedia has some dark corners they're going to be far better than the web at large
* Also a chance to assess likelihood of link breaking and if caching strategies like the above are worth it (probably not in the case of our own images).

**Comment by hyanwong** [comment]
> It might also be worth insisting that images, instead of being arbitrary URLs, are higher level constructs, e.g:

Yes, I think this is a good idea.

**Comment by hyanwong** [comment]
We can also store BLOBs in one of our databases if we like, but there would be a fair bit of plumbing there, I think.

**Comment by lentinj** [comment]
> It might also be worth insisting that images, instead of being arbitrary URLs, are higher level constructs...

For video this is essential, since the embed HTML for vimeo != youtube != HTTP-served mp4. So continuing this pattern would make sense.

**Comment by lentinj** [comment]
The start of this now exists here:

https://github.com/OneZoom/OZtree/commit/8916408662c6c96ed0e0c120c2ba4b94fdeeb718

We're just formatting HTML at this stage, but it's easy to imagine how the wikimedia commons processor could:

* If entry doesn't already exist in DB table...
  * Fetch the image metadata, insert into DB
  * Fetch the image itself, park it in a web-accessible directory
* Return HTML that includes attribution, and uses our copy of the image, rather than hot-linking wikimedia

Is this what we want to do, in this case and/or generally?

**Comment by lentinj** [comment]
A tourstop should wait for a video to end before advancing. But what if I leave a video paused? I think that should be equivalent to pausing the tourstop, and we should wait until the video actually reaches it's end.

**Comment by lentinj** [comment]
Also, to move off after a video has finished playing, you have to specify a ``wait`` period. Otherwise we wait for the user to press next as well as waiting for the video to finish. A wait period is also useful to paper over the time between a tourstop appearing and the video autoplaying (or never, if we're not allowed to autoplay yet). It feels like an annoying thing that will get forgotten.

The youtube tour plugin could force a wait period for any stop that doesn't have one, but the TourStops instances are already initialised by this point, so we'd have to do some jiggling about to change the value. And it's not entirely impossible that a tour author wants the user to watch the video, do something else, click next.

Finally, the name "wait" is problematic since it's a reserved word, so we can't have a column with that name. Instead I called it "stop_wait" in the model, and made an attempt to use Web2Py's ``rname`` to paper around the difference. However, this isn't working. Any objections to instead of working this out, just rename it to stop_wait everywhere and be done with it?

**Comment by jrosindell** [comment]
Great thoughts in the above.... 

I think an attribution URL might need to have two components a URL of who to credit for the content and a URL that shows the content is CC (unless that page is changed).  I agree there is no point in saving the content of that page as proof as it'll probably never come up.  

I can see a use case (down the line before third party tours are implemented) for a script that keeps local copies of third party assets needed for keeping tours running. Not necessarily to be used at all but just to be kept just in case some disaster knocks out a lot of our content.  A similar part of such a future project could test the URLs of third party assets and provide admin warnings in case tours are not working properly.

Overall, from reading the above, I guess I'm not so keen on the BLOB or data: URI in JSON solutions because of their associated performance issues. I may be wrong but it feels like hacking a database structure to do things that it wasn't really designed to do.

So, in conclusion, I'm agreeing with the higher level constructs for assets but with the added points about further data on attribution and the possible future scripts to protect ourselves against third part data failures.  I guess a bit part of the benefits here is that we can only allow certain types of constructs so a URL to some dodgy place we've not vetted will never work whereas a URL to anywhere in the domain of a vetted third party website would automatically be fine.

**Comment by jrosindell** [comment]
> We're just formatting HTML at this stage, but it's easy to imagine how the wikimedia commons processor could:
> 
> * If entry doesn't already exist in DB table...
>   
>   * Fetch the image metadata, insert into DB
>   * Fetch the image itself, park it in a web-accessible directory
> * Return HTML that includes attribution, and uses our copy of the image, rather than hot-linking wikimedia
> 
> Is this what we want to do, in this case and/or generally?

Brilliant!

**Comment by jrosindell** [comment]
I think the existing image handling gubbins is designed to use EOL images that have ratings and crop points to a square. I think it's a separate (but very useful) project to provide an interface that enables images of a species from wikipedia to be grabbed in and cropped via a user-assisted pipeline.  I'm not sure how much that would be able to use code from the downloading script being proposed above or how much work it would be to build.  I think it would pay quite big dividends though as we've been asked to improve the quality of our imagery by individual sponsors and funders and currently Yan does a lot of this by hand.  We've also been offered sets of brilliant images by third parties that we're not currently using.  Perhaps we could scope out a project around this to be done separately (maybe or maybe not as part of tours). It would be a good type of thing that could be assigned to trusted volunteers and would make a big difference to the quality of the tree.

**Comment by lentinj** [comment]
> I think an attribution URL might need to have two components...

Yes, but that'll be up to the plugin to work this out. For our wikipedia commons case we need to generate the equivalent of the HTML code in "Use this file on the web" on this page https://commons.wikimedia.org/wiki/File:Rose_of_Jericho.gif.

> I can see a use case (down the line before third party tours are implemented) for a script that keeps local copies of third party assets needed for keeping tours running. Not necessarily to be used at all but just to be kept just in case some disaster knocks out a lot of our content.

I think you either build it into how the system works, or you accept they may disappear. A script that may never get used will probably not work when it comes to finally using it anger...

 > I guess I'm not so keen on the BLOB or data: URI in JSON solutions

Nah, they're not ideal for the sorts of things we're talking about. data: uris are a more appropriate tool for branding / logos though, where we're only talking about a few KB at most.

>  I think it's a separate (but very useful) project to provide an interface that enables images of a species from wikipedia to be grabbed in and cropped via a user-assisted pipeline.

Such an interface could certainly make use of this, especially if there's already a tour creator GUI component to manage image selection.

**Comment by jrosindell** [comment]
All sounds good and I think it means this aspect of the work is shelved for the time being but just taken into account when specifying the way the tours link to assets required.

**Comment by jrosindell** [comment]
I've opened [#741](https://github.com/OneZoom/OZtree/issues/741) with things that should probably have been here. In any case I think there is still a little tidying up here if we are to have sufficient flexibility over tours media.

**Comment by lentinj** [comment]
Closing this, since I don't think it's necessary along with the other 2 issues, which are a bit more targeted.
